framework/src/media/utils/MediaUtils.cpp

```diff
@@ -43,6 +150,393 @@ static const std::string MP4_MIME_TYPE = "audio/mp4";
+		uint16_t lenSection = (SECTION_HEAD_BYTES + SECTION_LENGTH(ptrSection));
+		if (lenSection > (lenPayload - 1 - u8PointerField)) {
+			// portion of section in this packet
+			return UTILS_ERROR_NOT_SUPPORTED;
+		}
+		// TODO: check CRC32
+		// section complete (whole section contained in one TS packet)
+		*section = ptrSection;
+		*sectlen = lenSection;
+		return UTILS_ERROR_NONE;
+	} else {
+		// section appending
+		return UTILS_ERROR_NOT_SUPPORTED;
+	}
+}
+
+// Get the information (number of programs) in the PAT.
+static int ts_psi_get_pat_info(ts_demux_t *demux, psi_pat_info_t *info)
+{
+	psi_section_t *patSect = &demux->pat.section;
+	info->tsId = (patSect->data[3] << 8) + patSect->data[4];
+	// take off 8bytes header and 4bytes CRC, one program consists of 4bytes.
+	info->numOfPrgs = (patSect->size - 12) >> 2;
+	return UTILS_ERROR_NONE;
+}
+
+// Get the program number and PID from PAT at the given program index.
+static int ts_psi_get_program(ts_demux_t *demux, uint16_t index, uint16_t *prgNum, uint16_t *pmtPid)
+{
+	psi_section_t *patSect = &demux->pat.section;
+	if (index >= ((patSect->size - 12) >> 2)) {
+		return UTILS_ERROR_BAD_PARAMETER;
+	}
+	// 8bytes header and 4bytes CRC, loop { 2bytes program number + 2bytes pmt pid }
+	*prgNum = (patSect->data[8 + index * 4] << 8) + patSect->data[9 + index * 4];
+	*pmtPid = ((patSect->data[10 + index * 4] << 8) + patSect->data[11 + index * 4]) & 0x1FFF;
+	return UTILS_ERROR_NONE;
+}
+
+// Get the information (program number, number of streams, etc.) in the PMT.
+static int ts_psi_get_pmt_info(ts_demux_t *demux, psi_pmt_info_t *info)
+{
+	psi_section_t *pmtSect = &demux->pmt.section;
+	info->prgNum = (pmtSect->data[3] << 8) + pmtSect->data[4];
+	info->prgInfoLen = ((pmtSect->data[10] << 8) + pmtSect->data[11]) & 0x0FFF;
+	info->numOfStreams = demux->numOfStreams;
+	return UTILS_ERROR_NONE;
+}
+
+// Get the information (stream type, stream PID, etc.) of the PMT elementary stream.
+static int ts_psi_get_stream_info(ts_demux_t *demux, uint16_t index, psi_pmt_stream_info_t *info)
+{
+	psi_section_t *pmtSect = &demux->pmt.section;
+	if (index >= demux->numOfStreams) {
+		return UTILS_ERROR_BAD_PARAMETER;
+	}
+	uint16_t offset = demux->streamOffset[index];
+	info->type = pmtSect->data[offset];
+	info->pid = ((pmtSect->data[offset + 1] << 8) + pmtSect->data[offset + 2]) & 0x1FFF;
+	info->esInfoLen = ((pmtSect->data[offset + 3] << 8) + pmtSect->data[offset + 4]) & 0x0FFF;
+	return UTILS_ERROR_NONE;
+}
+
+// Calculate the number of streams in PMT and map each stream offset
+static void map_pmt_streams(ts_demux_t *demux)
+{
+	psi_section_t *pmtSect = &demux->pmt.section;
+	// 12bytes header + program info length
+	demux->streamOffset[0] = 12 + PROGRAM_INFO_LENGTH(pmtSect->data);
+	int i = 0;
+	while (demux->streamOffset[i] < pmtSect->size - CRC32_BYTES) {
+		// loop until stream offset is on the first CRC byte
+		if (i == PMT_STREAMS_MAXIMUM) {
```

Comment:
the logic is a little confusing here, you will first check the streamOffset[MIXIMUM] value, and then found it been MAXIMUM and break. how about put this checking under the value assignment, then you may break ealier

Reply:
I refined the logic to improve readability.